### ISSUE TITLE
Options not always reflected in older build directories

### ISSUE THREAD

**Issue by codido** [mesonbuild/meson#1344]
When meson_options.txt is created from scratch, invoking ninja in an older build directory doesn't seem to re-generate build.ninja. An easy workaround is to touch meson.build, and re-run ninja.

Also, if meson_options.txt is modified, it isn't reflected in mesonconf until ninja is invoked again.

**Comment by nirbheek** [comment]
I'm not sure if this is a bug because if you create a `meson_options.txt` from scratch, it's a useless file because you're not actually using the options defined in it anywhere in your build files. And as you say, when you do use the options in your build files, Ninja will regenerate itself.

**Comment by codido** [comment]
Right, though it does depend on the order of the user's actions - for example, if you try to use an option without creating meson_options.txt, build the project, realize you're actually missing it, add it and then rebuild it still wont work.
Also, even if you don't actually use any of options, you might still like to verify the options look sane in mesonconf before actually dealing with them.  But yeah, hardly a huge issue ;)

I think the second issue above, the need to re-run ninja before the new options are reflected in mesonconf, is probably slightly more important.

**Comment by nirbheek** [comment]
`mesonconf` operates only on private data files in the build directory and it doesn't look at the build files in the source tree at all. That's why it takes the build directory as an argument! It can only give accurate information when used with an up-to-date build directory.

Perhaps we should stat all build files and `meson_options.txt` and print a warning if the private data files are older.

**Comment by codido** [comment]
That was my first shot at this as well, but any reason it shouldn't just update the build directory options if it knows it should anyway?

**Comment by nirbheek** [comment]
Because that would require a regen, and we never automatically regen via `meson*` commands. The only time we do a regen is if you invoke `ninja`. Doing it in any other situation would be unexpected and can mess up people's build directories.

**Comment by codido** [comment]
Not sure why regenerating is such a bad idea - it will happen anyway on the next ninja invocation (for any target), wouldn't it?
But in any case, if that's a requirement, issuing a warning and perhaps bailing out seems like the appropriate way to deal with this.

**Comment by nirbheek** [comment]
First, calling `mesonconf` has always been a read-only operation. Making it regen will suddenly make it a sometimes-write operation, and the user will not necessarily expect it.

Second, regenerating outside of `ninja` can be a bad idea because sometimes people set environment variables that are used in the configure/compile steps which Meson does not know about, and doing an automatic regen will break people's build directories requiring a fresh one.

Third, just because the source directory has changes doesn't mean the user wants to regenerate the build directory. What if they just want to see the options that were used while creating that build directory? This is a very common use-case if you are building different branches of the same source tree in two different build directories.

**Comment by codido** [comment]
Well, correct me if I'm wrong, but at least to some extent mesonconf isn't
really read-only - you can set options after all.
I was under the impression that the options are kept separately from the
actual build.ninja file, so keeping the options up-to-date could be done
without re-generating build.ninja. But I might be missing some use cases
here.

It's clearly a matter of opinion, but as a newcomer I found that a bit
counter-intuitive. In order to set a new option, you need to invoke ninja
(explicitly stating the target as build.ninja), run mesonconf and then
re-run ninja again. At least for kbuild, when a kconfig file is changed you
don't really need to rebuild the kernel just so the change would be
reflected in menuconfig.

But again, if there are use cases that might be broken but this, a warning
is probably better than nothing at all.

On Tue, Jan 31, 2017 at 11:17 PM, Nirbheek Chauhan <notifications@github.com
